Complete the doxygen documentation

nondetalle opened the pull request:
I took sometime to manually write down all the missing documentation following the warnings produced by Doxygen. 
This allows us to solve the last failing check in the CI pipline. 

On top of Doxygen warnings, I did some maintenance to default comments for constructor, destructor and `GetTypeId` functions following the indications of [the ns-3 manual](https://www.nsnam.org/docs/release/3.41/manual/html/documentation.html) on the subject. Some further maintenance was to make sure that all sentences start with a capital letter and end with a column. I did not fully review existing documentation, but I think this is currently enough.

**This pull request exclusively concerns code documentation.** With the exception of a couple of function reordering in headers, no changes were applied to the running part of the code. That said, this manual review lead to me finding many problems: for that I left `\todo` tags in the Doxygen comments.  

I tried to keep the commit history as clean as possible for review purposes. With few exceptions, commit comments starting with _document_ concern one single header file among the ones affected by warnings. At the end, we can squash them all together.

nondetalle committed:
document GetTypeId

nondetalle committed:
document PeriodicSender

nondetalle committed:
add missing compound class/struct documentation in group lorawan

nondetalle committed:
document LorawanMac

nondetalle committed:
document LorawanMacHeader

nondetalle committed:
expand compound doc of HexGridPositionAllocator

nondetalle committed:
Add docs todo on unimplemented tests

nondetalle committed:
document LoraPacketTracker

nondetalle committed:
document EndDeviceStatus

nondetalle committed:
document MacCommand

nondetalle committed:
document lorawan-test-suite.cc

nondetalle committed:
document AdrComponent

nondetalle committed:
document EndDeviceLorawanMac

nondetalle committed:
document LoraHelper

nondetalle committed:
document LoraDeviceAddress

nondetalle committed:
document LoraFrameHeader

nondetalle committed:
document LoraInterferenceHelper

nondetalle committed:
document NetworkStatus

nondetalle committed:
document LorawanMacHelper

nondetalle committed:
document NetworkServerHelper

nondetalle committed:
document PeriodicSenderHelper

nondetalle committed:
document NetworkServer

nondetalle committed:
document network-server-test-suite.cc

nondetalle committed:
document aloha-throughput.cc

nondetalle committed:
document CorrelatedShadowingPropagationLossModel

nondetalle committed:
document complete-network-example.cc

nondetalle committed:
document GatewayStatus

nondetalle committed:
document LogicalLoraChannel

nondetalle committed:
document OneShotSenderHelper

nondetalle committed:
document HexGridPositionAllocator

nondetalle committed:
document NetworkScheduler

nondetalle committed:
document frame-counter-update.cc

nondetalle committed:
document ForwarderHelper

nondetalle committed:
document GatewayLoraPhy

nondetalle committed:
Remove obsolete mentions to CallBackTraceSource

nondetalle committed:
document NetworkController

nondetalle committed:
Document EndDeviceLoraPhy

nondetalle committed:
document test/utilities.h

nondetalle committed:
document LoraNetDevice

nondetalle committed:
document LoraPhy

nondetalle committed:
document NetworkControllerComponents

nondetalle committed:
document adr-example.cc

nondetalle committed:
document LoraPhyHelper

nondetalle committed:
document BuildingPropagationLoss

nondetalle committed:
document ClassAEndDeviceLorawanMac

nondetalle committed:
document Forwarder

nondetalle committed:
document GatewayLorawanMac

nondetalle committed:
document LoraRadioEnergyModel

nondetalle committed:
document LoraTag

nondetalle committed:
document LoraTxCurrentModel

nondetalle committed:
document OneShotSender

nondetalle committed:
Minor fixes to documentation

nondetalle committed:
Remove optionality of Doxygen check from CI

nondetalle committed:
Fix unused comments causing accidental Doxygen warnings

nondetalle committed:
Minor corrections and improvements

nondetalle committed:
Use suggested documentation for GetTypeId

nondetalle committed:
Use suggested documentation for destructors

nondetalle committed:
Use suggested documentation for default constructors

nondetalle committed:
Remove \brief as suggested by ns-3 manual

nondetalle committed:
Grammar: capital letter at start of sentences and column at end

nondetalle committed:
Fix CI

nondetalle committed:
Fix doxygen rendering of ASCII diagram in CorrelatedShadowingPropagationLossModel

pagmatt commented on examples/complete-network-example.cc:
```diff
@@ -55,18 +55,19 @@ using namespace lorawan;
-double simulationTime = 600;
+int nDevices = 200;          //!< Number of end device nodes to create
+int nGateways = 1;           //!< Number of gateway nodes to create
+double radius = 6400;        //!< Radius (m) of the deplyoment
```
Suggest `radiusMeter` and `simulationTimeSeconds`  to make this even more clear throughout the example. Applies to most examples

pagmatt commented on examples/complete-network-example.cc:
```diff
@@ -55,18 +55,19 @@ using namespace lorawan;
 
 // Output control
-bool print = true;
+bool print = true; //!< Whether to print building information
```
`printBuildingInfo` ?

pagmatt commented on examples/frame-counter-update.cc:
```diff
@@ -91,6 +105,12 @@ OnMacPacketOutcome(uint8_t transmissions, bool successful, Time firstAttempt, Pt
+ * Set the position of an end device either in range or out of range.
+ *
+ * \param endDevice A Ptr to the Node of the end device.
+ * \param inRange Whether to put the end device in range or out of range.
```
```suggestion
 * \param inRange Whether to set the end device in range or out of range.
```

pagmatt commented on examples/frame-counter-update.cc:
```diff
@@ -91,6 +105,12 @@ OnMacPacketOutcome(uint8_t transmissions, bool successful, Time firstAttempt, Pt
 }
 
+/**
+ * Set the position of an end device either in range or out of range.
```
```suggestion
 * Set the position of an end device as either in range or out of range.
```

pagmatt commented on helper/lora-helper.h:
```diff
@@ -37,39 +37,38 @@ namespace lorawan
-     * method.
+     * \param phyHelper The PHY helper to create PHY objects.
+     * \param macHelper The MAC helper to create MAC objects.
+     * \param c The set of nodes on which a lora device must be created.
```
```suggestion
     * \param c The set of nodes on which a lora device will be installed.
```

pagmatt commented on helper/lora-helper.h:
```diff
@@ -37,39 +37,38 @@ namespace lorawan
-     * method.
+     * \param phyHelper The PHY helper to create PHY objects.
+     * \param macHelper The MAC helper to create MAC objects.
+     * \param node The node on which a lora device must be created.
```
```suggestion
     * \param node The node on which a lora device will be installed.
```

pagmatt commented on helper/lora-helper.h:
```diff
@@ -98,28 +106,65 @@ class LoraHelper
                                               Time interval);
 
+    /**
+     * Print PHY-level performance at every gateway in the container since last
```
```suggestion
     * Print the PHY-level performance of every gateway in the container since the last
```

pagmatt commented on helper/lora-helper.h:
```diff
@@ -98,28 +106,65 @@ class LoraHelper
-
-    time_t m_oldtime;
+    LoraPacketTracker* m_packetTracker = nullptr; //!< Pointer to the Packet Tracker object
+    time_t m_oldtime;                             //!< Real time of the last simulation time print
```
Real 
```suggestion
    time_t m_oldtime;                             //!< Real (i.e., physical) time of the last simulation time print
```

pagmatt commented on helper/lora-packet-tracker.h:
```diff
@@ -41,127 +41,247 @@ enum PhyPacketOutcome
+/**
+ * \ingroup lorawan
+ *
+ * Stores uplink packet metrics from the PHY layer of sender/receivers
```
```suggestion
 * Stores PHY layer uplink packet metrics of sender/receivers
```

pagmatt commented on helper/lora-packet-tracker.h:
```diff
@@ -41,127 +41,247 @@ enum PhyPacketOutcome
+/**
+ * \ingroup lorawan
+ *
+ * Stores packet metrics from the MAC layer of sender/receivers
```
```suggestion
 * Stores MAC layer packet metrics of sender/receivers
```

pagmatt commented on helper/lora-packet-tracker.h:
```diff
@@ -41,127 +41,247 @@ enum PhyPacketOutcome
+/**
+ * \ingroup lorawan
+ *
+ * Stores metrics of the packet retransmission process optionally enabled in the MAC layer of
```
```suggestion
 * Stores (optionally enabled) MAC layer packet retransmission process metrics of
```

pagmatt commented on helper/lora-radio-energy-model-helper.h:
```diff
@@ -29,51 +29,44 @@ namespace lorawan
+ * \ingroup lorawan
  *
- * This installer installs LoraRadioEnergyModel for only LoraNetDevice objects.
+ * Assign LoraRadioEnergyModel to devices.
```
```suggestion
 * Installs LoraRadioEnergyModel on devices.
```

pagmatt commented on helper/lorawan-mac-helper.h:
```diff
@@ -64,61 +70,118 @@ class LorawanMacHelper
+     *
+     * The DR of each device is set to the maximum possible for its transmissions to be correctly
+     * received by the gateway measuring the best RSSI from the device, mimicking the DR
+     * maximization part of the default online LoRaWAN ADR algorithm. Mind that a single RSSI
```
```suggestion
     * maximization part of the default online LoRaWAN ADR algorithm. Please note that a single RSSI
```

pagmatt commented on helper/lorawan-mac-helper.h:
```diff
@@ -64,61 +70,118 @@ class LorawanMacHelper
+     * received by the gateway measuring the best RSSI from the device, mimicking the DR
+     * maximization part of the default online LoRaWAN ADR algorithm. Mind that a single RSSI
+     * measurement between each device and gateway pair is taken for the DR assignment, so the
+     * assignment may be suboptimal in scenarios with a stochastical channel.
```
```suggestion
     * assignment may be suboptimal in scenarios with a time-varying channel.
```

pagmatt commented on helper/lorawan-mac-helper.h:
```diff
@@ -64,61 +70,118 @@ class LorawanMacHelper
+     * v[4] -> number of devices using DR1 \n
+     * v[5] -> number of devices using DR0 \n
+     *
+     * \todo This function will always cause an array out of bound access error because at one point
```
This seems quite problematic, I would either fix this before releasing the module or mark this function as deprecated/unavailable until we fix it

pagmatt commented on helper/lorawan-mac-helper.h:
```diff
@@ -127,53 +190,71 @@ class LorawanMacHelper
     void ApplyCommonAlohaConfigurations(Ptr<LorawanMac> lorawanMac) const;
 
-    ObjectFactory m_mac;
+    ObjectFactory m_mac;                       //!< Mac layer object factory
```
```suggestion
    ObjectFactory m_mac;                       //!< MAC layer object factory
```

pagmatt commented on helper/network-server-helper.h:
```diff
@@ -38,58 +38,99 @@ namespace lorawan
+     * Create one lorawan network server application on each of the Nodes in the
+     * NodeContainer.
+     *
+     * \param c The nodes on which to create the Applications. The nodes
```
```suggestion
     * \param c The nodes on which to install the Applications. The nodes
```

pagmatt commented on helper/network-server-helper.h:
```diff
@@ -38,58 +38,99 @@ namespace lorawan
+    /**
+     * Create one lorawan network server application on the Node.
+     *
+     * \param node The node on which to create the Application.
```
```suggestion
     * \param node The node on which to install the Application.
```

pagmatt commented on model/adr-component.h:
```diff
@@ -62,60 +68,123 @@ class AdrComponent : public NetworkControllerComponent
     double RxPowerToSNR(double transmissionPower) const;
 
+    /**
+     * Get the min RSSI (dBm) among gateway receptions of a same transmission.
```
```suggestion
     * Get the min RSSI (dBm) among gateways receiving the same transmission.
```

pagmatt commented on model/adr-component.h:
```diff
@@ -62,60 +68,123 @@ class AdrComponent : public NetworkControllerComponent
     double GetMinTxFromGateways(EndDeviceStatus::GatewayList gwList);
-
+    /**
+     * Get the max RSSI (dBm) among gateway receptions of a same transmission.
```
```suggestion
     * Get the max RSSI (dBm) among gateways receiving the same transmission.
```

pagmatt commented on model/adr-component.h:
```diff
@@ -62,60 +68,123 @@ class AdrComponent : public NetworkControllerComponent
     double GetMaxTxFromGateways(EndDeviceStatus::GatewayList gwList);
-
+    /**
+     * Get the average RSSI (dBm) of gateway receptions of a same transmission.
```
```suggestion
     * Get the average RSSI (dBm) of gateways receiving the same transmission.
```

pagmatt commented on model/building-penetration-loss.h:
```diff
@@ -56,37 +57,37 @@ class BuildingPenetrationLoss : public PropagationLossModel
      * Generate a random p value.
      * The distribution of the returned value is as specified in TR 45.820.
-     * \returns A value in the 0-3 range.
+     * \return A value in the 0-3 range.
```
```suggestion
     * \return A random value in the 0-3 range.
```

pagmatt commented on model/building-penetration-loss.h:
```diff
@@ -56,37 +57,37 @@ class BuildingPenetrationLoss : public PropagationLossModel
      * Get a value to compute the wall loss.
      * The distribution of the returned value is as specified in TR 45.820.
-     * \returns A value in the 0-2 range.
+     * \return A value in the 0-2 range.
```
```suggestion
     * \return A random value in the 0-2 range.
```

pagmatt commented on model/correlated-shadowing-propagation-loss-model.h:
```diff
@@ -30,53 +30,88 @@ class MobilityModel;
+/**
+ * \ingroup lorawan
+ *
+ * Propagation loss model for correlated shadowing in a city
```
```suggestion
 * Propagation loss model for spatially correlated shadowing in a city
```

pagmatt commented on model/end-device-lorawan-mac.h:
```diff
@@ -103,6 +110,8 @@ class EndDeviceLorawanMac : public LorawanMac
      *
      * This function handles opening of the first receive window.
+     *
+     * \param packet The packet that just finished transmission.
```
```suggestion
     * \param packet The packet that has just been transmitted.
```

pagmatt commented on model/end-device-lorawan-mac.h:
```diff
@@ -124,18 +133,24 @@ class EndDeviceLorawanMac : public LorawanMac
 
     /**
-     * Set the maximum number of transmissions allowed.
+     * Get the default number of unacknowledged redundant transmissions of each packet. If,
```
Is this the default or the max value?

pagmatt commented on model/end-device-lorawan-mac.h:
```diff
@@ -176,28 +191,28 @@ class EndDeviceLorawanMac : public LorawanMac
+    ///**
+    // * Set a value for the RX1DROffset parameter.
+    // *
+    // * This value decides the offset to use when deciding the DataRate of the
```
```suggestion
    // * This value determines the offset to use when deciding the DataRate of the
```

pagmatt commented on model/end-device-status.h:
```diff
@@ -94,17 +92,10 @@ class EndDeviceStatus : public Object
-
-        // Whether or not this device needs a reply
-        bool needsReply = false;
+        LorawanMacHeader macHeader;  //!< The Mac Header to attach to the reply packet.
```
```suggestion
        LorawanMacHeader macHeader;  //!< The MAC Header to attach to the reply packet.
```

pagmatt commented on model/gateway-lora-phy.h:
```diff
@@ -69,10 +75,21 @@ class GatewayLoraPhy : public LoraPhy
     bool IsTransmitting() override;
 
+    /**
+     * Check whether the GatewayLoraPhy is currently listening to a frequency.
```
```suggestion
     * Check whether the GatewayLoraPhy is currently listening to the specified frequency.
```

pagmatt commented on model/gateway-lorawan-mac.h:
```diff
@@ -52,8 +65,9 @@ class GatewayLorawanMac : public LorawanMac
 
     /**
-     * Return the next time at which we will be able to transmit.
+     * Return the next time at which we will be able to transmit on a frequency.
```
```suggestion
     * Return the next time at which we will be able to transmit on the specified frequency.
```

pagmatt commented on model/lora-net-device.h:
```diff
@@ -46,37 +55,40 @@ class LorawanMac;
      * Set which LorawanMac instance is linked to this device.
      *
-     * \param mac the mac layer to use.
+     * \param mac The mac layer to use.
```
```suggestion
     * \param mac The MAC layer to use.
```

pagmatt replied:
This seems more in line with the remainder of the doc to me

pagmatt commented on model/lora-net-device.h:
```diff
@@ -46,37 +55,40 @@ class LorawanMac;
      * Set which LoraPhy instance is linked to this device.
      *
-     * \param phy the phy layer to use.
+     * \param phy The phy layer to use.
```
```suggestion
     * \param phy The PHY layer to use.
```

pagmatt commented on model/lora-net-device.h:
```diff
@@ -46,37 +55,40 @@ class LorawanMac;
      * Get the LorawanMac instance that is linked to this NetDevice.
      *
-     * \return the mac we are currently using.
+     * \return The mac we are currently using.
```
```suggestion
     * \return The MAC we are currently using.
```

pagmatt commented on model/lora-net-device.h:
```diff
@@ -46,37 +55,40 @@ class LorawanMac;
      * Get the LoraPhy instance that is linked to this NetDevice.
      *
-     * \return the phy we are currently using.
+     * \return The phy we are currently using.
```
```suggestion
     * \return The PHY we are currently using.
```

pagmatt commented on model/lora-radio-energy-model.h:
```diff
@@ -160,64 +160,64 @@ class LoraRadioEnergyModel : public DeviceEnergyModel
+     * Gets idle current.
      *
-     * \returns idle current of the lora device.
+     * \return Idle current of the lora device.
```
Unit of measure?

pagmatt commented on model/lora-radio-energy-model.h:
```diff
@@ -160,64 +160,64 @@ class LoraRadioEnergyModel : public DeviceEnergyModel
+     * Sets idle current.
      *
-     * \param idleCurrentA the idle current
+     * \param idleCurrentA The idle current.
```
```suggestion
     * \param idleCurrentA The idle current [A].
```

pagmatt commented on model/lora-radio-energy-model.h:
```diff
@@ -160,64 +160,64 @@ class LoraRadioEnergyModel : public DeviceEnergyModel
+     * Gets transmit current.
      *
-     * \returns transmit current of the lora device.
+     * \return Transmit current of the lora device.
```
```suggestion
     * \return Transmit current of the lora device [A].
```

pagmatt commented on model/lora-radio-energy-model.h:
```diff
@@ -160,64 +160,64 @@ class LoraRadioEnergyModel : public DeviceEnergyModel
+     * Sets transmit current.
      *
-     * \param txCurrentA the transmit current
+     * \param txCurrentA The transmit current.
```
```suggestion
     * \param txCurrentA The transmit current [A].
`` `

pagmatt replied:
Same for similar docs in this class which lack the unit of measure

pagmatt commented on model/mac-command.h:
```diff
@@ -213,7 +219,15 @@ class LinkAdrReq : public MacCommand
+     * Constructor with given fields.
+     *
+     * \param dataRate The DataRate field to set.
+     * \param txPower The TXPower field to set.
```
unit of measure?

nondetalle replied:
Not here, these are serializable fields of the header for configurations specified in the LoRaWAN protocol. 
They will translate into modulation settings with physical units of measure for the PHY layer of end devices, but here the values (few bits) only represent them indirectly.

pagmatt replied:
Got it, How about specifying that this is a serialized value in the doc and basically what adding what you wrote here in the above comment?

nondetalle replied:
How about now?

pagmatt commented on model/mac-command.h:
```diff
@@ -258,36 +272,47 @@ class LinkAdrReq : public MacCommand
-    uint16_t m_channelMask;
-    uint8_t m_chMaskCntl;
-    uint8_t m_nbRep;
+    uint8_t m_dataRate;     //!< The DataRate field
```
unit of measure?

pagmatt commented on model/mac-command.h:
```diff
@@ -258,36 +272,47 @@ class LinkAdrReq : public MacCommand
-    uint8_t m_chMaskCntl;
-    uint8_t m_nbRep;
+    uint8_t m_dataRate;     //!< The DataRate field
+    uint8_t m_txPower;      //!< The TXPower field
```
unit of measure?

pagmatt replied:
Same as above

pagmatt commented on model/sub-band.h:
```diff
@@ -53,20 +60,18 @@ class SubBand : public Object
+    ///**
+    // * Get the last frequency of the subband.
+    // *
+    // * \return The lowest frequency of the SubBand.
```
unit measure

pagmatt commented on model/sub-band.h:
```diff
@@ -53,20 +60,18 @@ class SubBand : public Object
     /**
      * Get the lowest frequency of the SubBand.
      *
      * \return The lowest frequency of the SubBand.
```
unit measure

pagmatt commented on model/sub-band.h:
```diff
@@ -100,7 +105,7 @@ class SubBand : public Object
      * Return whether or not a frequency belongs to this SubBand.
      *
-     * \param frequency the frequency we want to test against the current subband
+     * \param frequency The frequency we want to test against the current subband.
```
unit of measure

pagmatt reviewed (commented):
Left a few comments (mostly minor) on this, I used suggestions whenever possible so that you can quickly apply the suggested changes (whenever you agree with them)

albertogara commented on examples/aloha-throughput.cc:
```diff
@@ -55,17 +55,24 @@ using namespace lorawan;
+/**
+ * Record the beginning of a transmission by an end device.
+ *
+ * \param packet A Ptr to the Packet sent.
```
A pointer to the packet sent.

albertogara commented on examples/aloha-throughput.cc:
```diff
@@ -55,17 +55,24 @@ using namespace lorawan;
+ * Record the beginning of a transmission by an end device.
+ *
+ * \param packet A Ptr to the Packet sent.
+ * \param systemId Node id of the sender end device.
```
Shouldn't this attribute be called nodeId or senderNodeId , instead of systemId then?

albertogara commented on examples/aloha-throughput.cc:
```diff
@@ -75,6 +82,12 @@ OnTransmissionCallback(Ptr<const Packet> packet, uint32_t systemId)
+/**
+ * Record the correct reception of a packet by a gateway.
+ *
+ * \param packet A Ptr to the packet received.
```
Same comment as above,  A pointer to the packet received

albertogara commented on examples/frame-counter-update.cc:
```diff
@@ -91,6 +105,12 @@ OnMacPacketOutcome(uint8_t transmissions, bool successful, Time firstAttempt, Pt
+/**
+ * Set the position of an end device either in range or out of range.
+ *
+ * \param endDevice A Ptr to the Node of the end device.
```
A pointer...

albertogara commented on helper/forwarder-helper.h:
```diff
@@ -37,26 +37,53 @@ namespace lorawan
+     * set with SetAttribute or other functions of this class.
+     *
+     * \param node The node on which a Forwarder will be installed.
+     * \return A Ptr to the applications installed.
```
A pointer

albertogara commented on helper/lora-helper.h:
```diff
@@ -85,11 +84,20 @@ class LoraHelper
+     * For each input device print the current position, data rate and transmission power settings.
+     *
+     * \param endDevices The devices to track.
+     * \param gateways \todo Unused parameter.
```
Please add the description for gateway too. You can also but in parenthesis that is only a placeholder.

albertogara commented on helper/lora-packet-tracker.h:
```diff
@@ -41,127 +41,247 @@ enum PhyPacketOutcome
+    Time sendTime;     //!< Timestamp of the pkt leaving MAC layer to go down the stack of sender
+    Time receivedTime; //!< Time of first reception \todo Parameter set to max and not used
+    std::map<int, Time> receptionTimes; //!< Timestamp of the pkt leaving MAC layer to go up the
+                                        // stack, mapped by receiver's node id
```
I think you also need to put the second line with  //!< for doxygen to recognize it (but I am not 100% sure)

nondetalle replied:
You are correct, I missed it during refactoring.

albertogara commented on helper/lora-packet-tracker.h:
```diff
@@ -41,127 +41,247 @@ enum PhyPacketOutcome
+     * the perspective of the specified gateway - the number of such packets well received, lost
+     * to interference, lost to unavailability of the gateway's reception paths, lost because under
+     * RSSI sensitivity, and lost due to concurrent downlink transmission of the gateway.
+     *
```
the perspective of the specified gateway - the number of such packets **correctly** received, lost
     * to interference, lost to unavailability of the gateway's reception paths, lost **for being** under
     * RSSI sensitivity **threshold**, and lost due to concurrent downlink transmission of the gateway

albertogara commented on helper/network-server-helper.h:
```diff
@@ -38,58 +38,99 @@ namespace lorawan
      * Enable (true) or disable (false) the ADR component in the Network
      * Server created by this helper.
+     *
+     * \param enableAdr Whether to enable adr in the Network Server.
```
What does Adr stand for? If possible add it in the description here.

albertogara commented on helper/network-server-helper.h:
```diff
@@ -38,58 +38,99 @@ namespace lorawan
 
-    ObjectFactory m_adrSupportFactory;
+    ObjectFactory m_factory;           //!< Factory to create the Network server application
+    NodeContainer m_gateways;          //!< Set of gateways to connect to this NS
```
Avoid using abbreviations or acronyms unless they are super well-known like RSSI.

I think is not the case for NS and ADR.

albertogara commented on model/adr-component.h:
```diff
@@ -62,60 +68,123 @@ class AdrComponent : public NetworkControllerComponent
     double GetMaxSNR(EndDeviceStatus::ReceivedPacketList packetList, int historyRange);
-
+    /**
+     * Get the average SNR of received packet history.
```
Get the average signal noise ratio (SNR) of the received packet history

albertogara commented on model/adr-component.h:
```diff
@@ -62,60 +68,123 @@ class AdrComponent : public NetworkControllerComponent
     double GetMinSNR(EndDeviceStatus::ReceivedPacketList packetList, int historyRange);
-
+    /**
+     * Get the max SNR of received packet history.
```
Get the max signal noise ratio (SNR) of the receive packet history.

albertogara commented on model/class-a-end-device-lorawan-mac.h:
```diff
@@ -111,13 +112,14 @@ class ClassAEndDeviceLorawanMac : public EndDeviceLorawanMac
      * Get the Data Rate that will be used in the first receive window.
      *
-     * \return The Data Rate
+     * \return The Data Rate.
```
The data rate.

albertogara commented on model/class-a-end-device-lorawan-mac.h:
```diff
@@ -131,21 +133,21 @@ class ClassAEndDeviceLorawanMac : public EndDeviceLorawanMac
      * Get the Data Rate that will be used in the second receive window.
      *
-     * \return The Data Rate
+     * \return The Data Rate.
```
The data rate.

albertogara commented on model/correlated-shadowing-propagation-loss-model.h:
```diff
@@ -107,20 +142,29 @@ class CorrelatedShadowingPropagationLossModel : public PropagationLossModel
+     * \todo Not implemented.
+     *
+     * \return The correlation distance as a double.
      */
```
if SetCorrelation() and GetCorrelation() are not implemented, there is no need to be declared or documented either.

albertogara commented on model/end-device-lorawan-mac.h:
```diff
@@ -349,57 +377,37 @@ class EndDeviceLorawanMac : public LorawanMac
+    bool m_enableDRAdapt; //!< Enable Data Rate adaptation during the retransmission procedure.
+    uint8_t
+        m_maxNumbTx; //!< Default number of unacknowledged redundant transmissions of each packet.
+    TracedValue<uint8_t> m_dataRate; //!< The DataRate this device is using to transmit.
```
The data rate

albertogara commented on model/end-device-lorawan-mac.h:
```diff
@@ -349,57 +377,37 @@ class EndDeviceLorawanMac : public LorawanMac
-     * The address of this device.
-     */
-    LoraDeviceAddress m_address;
+    bool m_enableDRAdapt; //!< Enable Data Rate adaptation during the retransmission procedure.
```
Enable data rate adaptation (ADR) during .....

albertogara commented on model/end-device-status.h:
```diff
@@ -197,20 +204,26 @@ class EndDeviceStatus : public Object
+     *
+     * \todo rename this to GetSecondReceiveWindowSpreadingFactor. No such thing as a SF offset for
+     * RX2 in the specifications. Also this function is unused.
      */
```
If this function is unused or serves no purpose, delete it instead.
Otherwise, rename it with the correct name and description.

albertogara commented on model/hex-grid-position-allocator.h:
```diff
@@ -25,46 +25,111 @@
 namespace ns3
 {
 
+/**
```
I think this is more fitting for documentation instead of API class description.
Consider moving it to documentation and placing a brief description here instead.

nondetalle replied:
The module documentation will need some rework as well in the future, currently it is not displaying correctly. For now I'd prefer to leave it as it is and add a \todo reminder.

nondetalle committed:
Apply PR review comments to examples
+ Clarify units of measures in examples' input variables

albertogara reviewed (commented):
I also left a brief comment about the documentation.

nondetalle committed:
Fix SetSpreadingFactorsGivenDistribution

nondetalle committed:
Apply PR review comments on helpers

nondetalle committed:
Apply PR review comments on model

nondetalle committed:
Refactor 'A Ptr' -> 'A pointer'
+ add the 'to' of Signal to Noise Ratio

nondetalle committed:
Expand ADR acronym

nondetalle commented:
Thank you @pagmatt and @albertogara for the enormous effort of reviewing the API documentation.
Let me know if you need more time.

I fixed most of your suggestions, leaving some yet to do or open for further discussion, and rebased onto `develop`.
I am trying to generalize your comments to the whole API documentation, because many of your remarks are also present in other API comments I had not touched with the initial changes intended to solve the Doxygen warnings. Tomorrow I'll close up what's left to do. 

You may have noticed that we are having problems with the CI pipeline after upgrading to ns-3.41. It appears that one test suite will always fail if we reuse cached build files, while it will work when the cache is not used. I have not been able to reproduce the problem locally yet, but I'll try to fix it before release.

Best,
Alessandro

nondetalle committed:
Expand DR acronym

nondetalle committed:
Expand GW acronym and uniform gateway keyword

nondetalle committed:
Expand ED acronym and uniform end device keyword

nondetalle committed:
Expand plural aconyms (EDs, GWs)

nondetalle committed:
Expand NS acronym and uniform network server keyword

nondetalle committed:
Expand BW acronym

nondetalle committed:
Expand SF acronym

nondetalle committed:
Fix errors in the CI pipeline introduced by ns-3.41
+ CI: only skip tests if they were successful on a past run with same ccache contents
+ Tests: compare packets with uid instead of each serialized byte

pagmatt commented:
> Thank you @pagmatt and @albertogara for the enormous effort of reviewing the API documentation. Let me know if you need more time.
> 
> I fixed most of your suggestions, leaving some yet to do or open for further discussion, and rebased onto `develop`. I am trying to generalize your comments to the whole API documentation, because many of your remarks are also present in other API comments I had not touched with the initial changes intended to solve the Doxygen warnings. Tomorrow I'll close up what's left to do.
> 
> You may have noticed that we are having problems with the CI pipeline after upgrading to ns-3.41. It appears that one test suite will always fail if we reuse cached build files, while it will work when the cache is not used. I have not been able to reproduce the problem locally yet, but I'll try to fix it before release.
> 
> Best, Alessandro

From my side, I honestly don't think I will have the time (surely not in the next few days) to go through the documentation again. So I suggest to simply finish discussing the few open points, then merge this and thus move forward with the updated release and finally have a fully-working CI pipeline.

nondetalle committed:
Improve LoRaWAN parameters description for LinkAdrReq

nondetalle committed:
Remove unused functions and improve \todo description

nondetalle committed:
Fix on previous commit

nondetalle committed:
Fix image display in module's .rst documentation

nondetalle commented:
Hello, if you agree we can merge the branch once the CI pipeline is done.
